@xen-orchestra/lite/src/libs/xen-api.ts

```diff
@@ -307,77 +307,57 @@ export default class XenApi {
   }
 
   get vm() {
-    type VmsRef =
-      | {
-          vmRef: XenApiVm["$ref"];
-          vmsRef?: undefined;
-        }
-      | {
-          vmRef?: undefined;
-          vmsRef: XenApiVm["$ref"][];
-        };
+    type VmRefs = XenApiVm["$ref"] | XenApiVm["$ref"][];
+    type VmRefsWithPowerState = Record<
+      XenApiVm["$ref"],
+      XenApiVm["power_state"]
+    >;
 
     return {
-      start: ({ vmRef, vmsRef }: VmsRef) => {
-        const _vmsRef = vmsRef ?? [vmRef];
-        return Promise.all(
-          _vmsRef.map((vmRef) => this._call("VM.start", [vmRef, false, false]))
-        );
-      },
-      startOn: ({ vmRef, vmsRef, hostRef }: VmsRef & { hostRef: string }) => {
-        const _vmsRef = vmsRef ?? [vmRef];
-        return Promise.all(
-          _vmsRef.map((vmRef) =>
+      start: (vmRefs: VmRefs) =>
+        Promise.all(
+          castArray(vmRefs).map((vmRef) =>
+            this._call("VM.start", [vmRef, false, false])
+          )
+        ),
+      startOn: (vmRefs: VmRefs, hostRef: XenApiHost["$ref"]) =>
+        Promise.all(
+          castArray(vmRefs).map((vmRef) =>
             this._call("VM.start_on", [vmRef, hostRef, false, false])
           )
-        );
-      },
-      pause: ({ vmRef, vmsRef }: VmsRef) => {
-        const _vmsRef = vmsRef ?? [vmRef];
+        ),
+      pause: (vmRefs: VmRefs) =>
+        Promise.all(
+          castArray(vmRefs).map((vmRef) => this._call("VM.pause", [vmRef]))
+        ),
+      suspend: (vmRefs: VmRefs) => {
         return Promise.all(
-          _vmsRef.map((vmRef) => this._call("VM.pause", [vmRef]))
+          castArray(vmRefs).map((vmRef) => this._call("VM.suspend", [vmRef]))
         );
       },
-      suspend: ({ vmRef, vmsRef }: VmsRef) => {
-        const _vmsRef = vmsRef ?? [vmRef];
-        return Promise.all(
-          _vmsRef.map((vmRef) => this._call("VM.suspend", [vmRef]))
-        );
-      },
-      resume: ({ vmRef, vmsRef }: VmsRef) => {
-        const _vmsRef = vmsRef ?? [vmRef];
-        const vmStore = useVmStore();
+      resume: (vmRefsWithPowerState: VmRefsWithPowerState) => {
+        const vmRefs = Object.keys(vmRefsWithPowerState);
+
         return Promise.all(
-          _vmsRef.map((ref) => {
-            const isSuspended =
-              vmStore.getRecord(ref).power_state === "Suspended";
-            return this._call(
-              `VM.${isSuspended ? "resume" : "unpause"}`,
-              isSuspended ? [ref, false, false] : [ref]
-            );
+          vmRefs.map((vmRef) => {
+            if (vmRefsWithPowerState[vmRef] === "Suspended") {
+              return this._call("VM.resume", [vmRef, false, false]);
+            }
+
+            return this._call("VM.unpause", [vmRef]);
```

Comment:
> Argument contains all the VMs to be resumed and we call either resume or unpause based on their current state.

Yes, but you don't know if the user actually selected VMs that were in the correct power state to be resumed or unpaused (ie suspended or paused). If they selected a running VM and you call `VM.unpause` on it, it will trigger a XAPI error.